tests/meshcop/joiner_session.cpp

```diff
@@ -0,0 +1,173 @@
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   The file implements the dtls session class to communicate with the joiner
+ */
+
+#include "joiner_session.hpp"
+#include "commissioner_utils.hpp"
+#include "agent/uris.hpp"
+#include "common/logging.hpp"
+#include "common/tlv.hpp"
+
+namespace ot {
+namespace BorderRouter {
+
+JoinerSession::JoinerSession(uint16_t aInternalServerPort, const char *aPskdAscii)
+    : mDtlsServer(Dtls::Server::Create(aInternalServerPort, JoinerSession::HandleSessionChange, this))
+    , mCoapAgent(Coap::Agent::Create(JoinerSession::SendCoap, this))
+    , mJoinerFinalizeHandler(OT_URI_PATH_JOINER_FINALIZE, HandleJoinerFinalize, this)
+    , mNeedAppendKek(false)
+{
+    mDtlsServer->SetPSK((const uint8_t *)aPskdAscii, strlen(aPskdAscii));
+    mDtlsServer->Start();
+    mCoapAgent->AddResource(mJoinerFinalizeHandler);
+}
+
+void JoinerSession::HandleSessionChange(Dtls::Session &aSession, Dtls::Session::State aState, void *aContext)
+{
+    JoinerSession *joinerSession = static_cast<JoinerSession *>(aContext);
+
+    switch (aState)
+    {
+    case Dtls::Session::kStateReady:
+        memcpy(joinerSession->mKek, aSession.GetKek(), sizeof(joinerSession->mKek));
+        aSession.SetDataHandler(JoinerSession::FeedCoap, joinerSession);
+        joinerSession->mDtlsSession = &aSession;
+        break;
+
+    case Dtls::Session::kStateClose:
+        break;
+
+    case Dtls::Session::kStateError:
+    case Dtls::Session::kStateEnd:
+        joinerSession->mDtlsSession = NULL;
+        break;
+    default:
+        break;
+    }
+}
+
+ssize_t JoinerSession::SendCoap(const uint8_t *aBuffer,
+                                uint16_t       aLength,
+                                const uint8_t *aIp6,
+                                uint16_t       aPort,
+                                void *         aContext)
+{
+    ssize_t        ret           = 0;
+    JoinerSession *joinerSession = static_cast<JoinerSession *>(aContext);
+
+    if (joinerSession->mDtlsSession)
+    {
+        ret = joinerSession->mDtlsSession->Write(aBuffer, aLength);
+    }
+    else
+    {
+        otbrLog(OTBR_LOG_INFO, "SendCoap: error NO SESSION");
+        ret = -1;
+    }
+
+    (void)aIp6;
+    (void)aPort;
+
+    return ret;
+}
+
+void JoinerSession::FeedCoap(const uint8_t *aBuffer, uint16_t aLength, void *aContext)
+{
+    JoinerSession *joinerSession = static_cast<JoinerSession *>(aContext);
+    joinerSession->mCoapAgent->Input(aBuffer, aLength, NULL, 1);
```

Comment:
Now it can be zero. In the old code, joiner session share same CoapAgent with commissioner, port number 1 was used to tell different traffic from each other.